quinn-proto/src/range_set/array_range_set.rs

```diff
@@ -0,0 +1,301 @@
+        // since clones (e.g. for storage in `SentFrames`) are rarely mutated
+        if self.0.is_inline() || self.0.len() > ARRAY_RANGE_SET_INLINE_CAPACITY {
+            return Self(self.0.clone());
+        }
+
+        let mut vec = TinyVec::new();
+        vec.extend_from_slice(self.0.as_slice());
+        Self(vec)
+    }
+}
+
+impl ArrayRangeSet {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    pub fn iter(&self) -> ArrayRangeSetIter<'_> {
+        ArrayRangeSetIter {
+            inner: &self.0,
+            front_idx: 0,
+            back_idx: self.0.len(),
+        }
+    }
+
+    pub fn elts(&self) -> ArrayRangeSetEltIter<'_> {
+        ArrayRangeSetEltIter {
+            inner: self.iter(),
+            start: 0,
+            end: 0,
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn contains(&self, x: u64) -> bool {
+        for range in self.0.iter() {
+            if range.start > x {
+                // We only get here if there was no prior range that contained x
+                return false;
+            } else if range.contains(&x) {
+                return true;
+            }
+        }
+        false
+    }
+
+    pub fn subtract(&mut self, other: &ArrayRangeSet) {
+        // TODO: This can potentially be made more efficient, since the we know
+        // individual ranges are not overlapping, and the next range must start
+        // where the last one finished
+        for range in &other.0 {
+            self.remove(range.clone());
+        }
+    }
+
+    pub fn insert_one(&mut self, x: u64) -> bool {
+        self.insert(x..x + 1)
+    }
+
+    pub fn insert(&mut self, mut x: Range<u64>) -> bool {
+        let mut result = false;
+
+        let mut idx = 0;
+        while idx != self.0.len() {
+            let range = &mut self.0[idx];
+
+            if range.start > x.end {
+                // The range is fully before this range and therefore not extensible.
+                // Add a new range to the left
+                self.0.insert(idx, x);
+                return true;
+            } else if range.start > x.start {
+                // The new range starts before this range but overlaps.
+                // Extend the current range to the left
+                // Note that we don't have to merge a potential left range, since
+                // this case would have been captured by merging the right range
+                // in the previous loop iteration
+                result = true;
+                range.start = x.start;
+                // Modify the range to the start of this range, so that we can
+                // afterwards check for extending to the right
+                #[allow(clippy::almost_swapped)]
+                {
+                    x.start = range.start;
+                }
+                x.end = x.end.max(range.end);
+
+                if x.start >= x.end {
+                    return result;
```

Comment:
We assign `result = true` above here; should this be simplified to `return true;` for clarity?

Reply:
Yes. Doesn't really matter.